### PR TITLE
Make `SampleCurve`/`UnevenSampleCurve` succeed at reflection

### PR DESCRIPTION
**Pull request by mweatherley**
(Note: #15434 implements something very similar to this for functional curve adaptors, which is why they aren't present in this PR.)

# Objective

Previously, there was basically no chance that the explicitly-interpolating sample curve structs from the `Curve` API would actually be `Reflect`. The reason for this is functional programming: the structs contain an explicit interpolation `I: Fn(&T, &T, f32) -> T` which, under typical circumstances, will never be `Reflect`, which prevents the derive from realistically succeeding. In fact, they won't be a lot of other things either, notably including both`Debug` and `TypePath`, which are also required for reflection to succeed. 

The goal of this PR is to weaken the implementations of reflection traits for these structs so that they can implement `Reflect` under reasonable circumstances. (Notably, they will still not be `FromReflect`, which is unavoidable.)

## Solution

The function fields are marked as `#[reflect(ignore)]`, and the derive macro for `Reflect` has `FromReflect` disabled. (This is not fully optimal, but we don't presently have any kind of "read-only" attribute for these fields.) Additionally, these structs receive custom `Debug` and `TypePath` implementations that display the function's (unstable!) type name instead of its value or type path (respectively). In the case of `TypePath`, this is a bit janky, but the instability of `type_name` won't generally present an issue for generics, which would have to be registered manually in the type registry anyway, which is impossible because the function type parameters cannot be named.

(And in general, the "blessed" route for such cases would generally involve manually monomorphizing the function parameter away, which also allows access to `FromReflect` etc. through very ordinary use of the derive macro.)

## Testing

Tests in the new `bevy_math::curve::sample_curves` module guarantee that these are actually `Reflect` under reasonable circumstances.

---

## Future changes

If and when function item types become `Default`, these types will need to receive custom `FromReflect` implementations that exploit it. Such a custom implementation would also be desirable if users start doing things like wrapping function items in `Default`/`FromReflect` wrappers that still implement a `Fn` trait. 

Additionally, if function types become nameable in user-space, the stance on `Debug`/`TypePath` may bear reexamination, since partial monomorphization through wrappers would make implementing reflect traits for function types potentially more viable.

### CODE REVIEW

File: crates/bevy_math/src/curve/sample_curves.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    }
+}
+
+/// A curve that is defined by neighbor interpolation over a set of evenly-spaced samples,
+/// interpolated automatically using [a particularly well-behaved interpolation].
+///
+/// [a particularly well-behaved interpolation]: StableInterpolate
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
```
Comment (MrGVSV) [inline crates/bevy_math/src/curve/sample_curves.rs:126]: Nit:

```suggestion
#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Debug))]
```

File: crates/bevy_math/src/curve/sample_curves.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+#[cfg(feature = "bevy_reflect")]
+use bevy_reflect::{utility::GenericTypePathCell, Reflect, TypePath};
+
+const THIS_MODULE: &str = "bevy_math::curve::sample_curves";
+const THIS_CRATE: &str = "bevy_math";
+
+/// A curve that is defined by explicit neighbor interpolation over a set of evenly-spaced samples.
+#[derive(Clone)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
```
Comment (MrGVSV) [inline crates/bevy_math/src/curve/sample_curves.rs:18]: Should we `#[reflect(Serialize, Deserialize)]`? Same goes for the other types.

File: crates/bevy_math/src/curve/sample_curves.rs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(Reflect),
+    reflect(where T: TypePath),
+    reflect(from_reflect = false, type_path = false),
+)]
```
Comment (MrGVSV) [inline crates/bevy_math/src/curve/sample_curves.rs:24]: Nit:
```suggestion
#[cfg_attr(
    feature = "bevy_reflect",
    derive(Reflect),
    reflect(where T: TypePath),
    reflect(from_reflect = false, type_path = false),
    reflect(Debug),
)]
```

File: crates/bevy_math/src/curve/sample_curves.rs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+/// A curve that is defined by interpolation over unevenly spaced samples with explicit
+/// interpolation.
+#[derive(Clone)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(Reflect),
+    reflect(where T: TypePath),
+    reflect(from_reflect = false, type_path = false),
```
Comment (MrGVSV) [inline crates/bevy_math/src/curve/sample_curves.rs:169]: Nit: 
```suggestion
    reflect(from_reflect = false, type_path = false),
    reflect(Debug),
```

File: crates/bevy_math/src/curve/sample_curves.rs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+    }
+}
+
+/// A curve that is defined by interpolation over unevenly spaced samples,
+/// interpolated automatically using [a particularly well-behaved interpolation].
+///
+/// [a particularly well-behaved interpolation]: StableInterpolate
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
```
Comment (MrGVSV) [inline crates/bevy_math/src/curve/sample_curves.rs:282]: Nit: 
```suggestion
#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Debug))]
```

**Review by MrGVSV — APPROVED** [review]
From a reflection perspective this looks good to me!

**Comment by MrGVSV** [comment]
> The reason for this is functional programming: the structs contain an explicit interpolation `I: Fn(&T, &T, f32) -> T` which, under typical circumstances, will never be `Reflect`, which prevents the derive from realistically succeeding.

I wonder if we could generate an anonymous-scoped type that wraps the `Fn(&T, &T, f32) -> T` and then uses remote reflection to power the reflection API for that function pointer. Or maybe even a generic one provided by `bevy_reflect`. 🤔

**Comment by MrGVSV** [comment]
> If and when function item types become `Default`, these types will need to receive custom `FromReflect` implementations that exploit it.

Not necessarily. Any fields marked `#[reflect(ignore)]` in a type not marked `#[reflect(from_reflect = false)]` will have a `Default` bound by default (you can also give it a custom default constructor if needed).

**Comment by mweatherley** [comment]
> Not necessarily. Any fields marked `#[reflect(ignore)]` in a type not marked `#[reflect(from_reflect = false)]` will have a `Default` bound by default (you can also give it a custom default constructor if needed).

That's fair, I was just thinking about `#[derive(Reflect)]`, but we can use `#[derive(FromReflect)]` separately for that.

> I wonder if we could generate an anonymous-scoped type that wraps the `Fn(&T, &T, f32) -> T` and then uses remote reflection to power the reflection API for that function pointer. Or maybe even a generic one provided by `bevy_reflect`. 🤔

I was also wondering about stuff of this nature. Maybe worth investigating!

**Comment by mweatherley** [comment]
Looks like the `#[reflect(Debug)]` stuff didn't work out; I'm content to look into it some other time, though.

File: crates/bevy_math/src/curve/sample_curves.rs (hunk 2 above)
Comment (MrGVSV) [reply crates/bevy_math/src/curve/sample_curves.rs:21]: Oh right, I always forget how much generic types mess up the type data registrations haha

Comment (mweatherley) [reply crates/bevy_math/src/curve/sample_curves.rs:21]: Yeah; on the other hand, they have a tendency to make them less useful as well, so 🤷

**Review by kristoff3r — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Merge conflicts are non-trivial; clean those up and I'll press the button.

**Pull request merged**